Replaced deprecated Jetpack::is_staging_site call

peterfabian opened the pull request:
Replaces #25665 .

### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #25664 .

### How to test the changes in this Pull Request:

1. Get a new JN instance
2. Build this package and upload it as a plugin to JN
3. Go through the OBW and 
- allow tracking, 
- activate Jetpack, 
4. you shouldn't see any error (ideally). Or at least not the error saying Notice: Jetpack::is_staging_site is deprecated since version jetpack-8.1! Use /Automattic/Jetpack/Status->is_staging_site instead.
4. Get another JN without Jetpack
5. Install Jetpack older than 8.1, e.g. 7.9 or 8.0
6. Verify that there are no error during OBW when activating Jetpack.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Fix - Replaced deprecated Jetpack::is_staging_site call.

peterfabian committed:
Replaced deprecated is_staging_site call.

rodrigoprimo commented on includes/class-wc-tracker.php:
```diff
@@ -111,9 +111,22 @@ private static function get_tracking_data() {
 
 		// Jetpack & WooCommerce Connect.
+		// Test if the site if a staging site based on the Jetpack criteria.
+		$is_jp_staging = false;
```
Maybe this code to decide which method to use to determine if the site is a staging site depending on the Jetpack available should be abstracted to its own method?

peterfabian replied:
Agreed, was thinking about this today as well.

rodrigoprimo commented on includes/class-wc-tracker.php:
```diff
@@ -111,9 +111,22 @@ private static function get_tracking_data() {
+		if ( class_exists( '\Automattic\Jetpack\Status' ) ) {
+			// Preferred way of checking with Jetpack 8.1+.
+			$jp_status = new \Automattic\Jetpack\Status();
+			if ( is_callable( '$jp_status->is_staging_site()' ) && $jp_status->is_staging_site() ) {
```
I don't think that `is_callable( '$jp_status->is_staging_site()' )` works. I believe it will always return false. To check if a given method of an object is callable you should pass an array where the first item is the object and the second item is the name of the method:

```php
is_callable( array( $jp_status, 'is_staging_site' ) )
```

peterfabian replied:
You're right of course. Thanks, fixed both items in https://github.com/woocommerce/woocommerce/pull/25670/commits/88b8c006dbd92e5f64ebc35ee5658c0fcac4c0a1, I believe.

rodrigoprimo reviewed (changes requested):
IMO the solution proposed in this PR is a good one, but there is a bug that needs to be fixed before merging. I also left a suggestion to change the organization of the code. More information in inline comments.

rodrigoprimo commented:
I forgot to mention that I was not able to reproduce this warning during the OBW but I didn't spend much time investigating why. To test the changes proposed in this PR, I changed the visibility of the method WC_Tracker::get_tracking_data() to public and called it directly.

peterfabian committed:
Extracted the test to a function and fixed is_callable call.

rodrigoprimo reviewed (approved):
Looks good to me now. Thanks for the fix.

kraftbj commented on includes/class-wc-tracker.php:
```diff
@@ -86,6 +86,30 @@ private static function get_last_send_time() {
+		if ( class_exists( '\Automattic\Jetpack\Status' ) ) {
+			// Preferred way of checking with Jetpack 8.1+.
+			$jp_status = new \Automattic\Jetpack\Status();
+			if ( is_callable( array( $jp_status, 'is_staging_site' ) ) && $jp_status->is_staging_site() ) {
```
There are versions of Jetpack with `\Automattic\Jetpack\Status` before `is_staging_site` was added to it and would still need to use the old fallback.

peterfabian replied:
Thanks, fixed in 3dcb3e0.

peterfabian committed:
Correct the fallback check logic for is_staging_site().

There are JP versions where \Automattic\Jetpack\Status exists, but does not
contain is_staging_site() function, so this commit should correctly handle
those cases as well.

peterfabian committed:
Merge branch 'master' into fix/25664-2

peterfabian commented:
Tested on JN with JP 8.2.3, 8.1, 8.0 and 7.9, it seems like it doesn't throw any deprecations anymore.

vedanshujain reviewed (approved):
LGTM! Like @rodrigoprimo  I was not able to repro on OBW so I called get_tracking_data directly to test.
